### PR TITLE
Better detection of ipykernel installer actions

### PR DESCRIPTION
**Pull request by DonJayamanne**

### CODE REVIEW

File: src/client/datascience/interactive-window/interactiveWindow.ts [hunk 1]
```diff
@@ -172,7 +172,7 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
     private async createKernelReadyPromise(): Promise<IKernel> {
         const editor = await this._editorReadyPromise;
         const controller = await this._controllerReadyPromise.promise;
-        initializeNotebookTelemetryBasedOnUserAction(editor.document.uri, controller!.connection);
+        initializeInteractiveOrNotebookTelemetryBasedOnUserAction(this.owner, controller!.connection);
```
Comment (DonJayamanne) [inline src/client/datascience/interactive-window/interactiveWindow.ts:175]: Was passing the wrong Uri

File: src/client/datascience/jupyter/kernels/helpers.ts [hunk 2]
```diff
@@ -770,29 +769,42 @@ export async function sendTelemetryForPythonKernelExecutable(
             match: match ? 'true' : 'false',
             kernelConnectionType: kernelConnection.kind
         });
-        trackKernelResourceInformation(Uri.file(file), { interpreterMatchesKernel: match });
+        trackKernelResourceInformation(resource, { interpreterMatchesKernel: match });
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/helpers.ts:772]: Passing the Uri instead,

File: src/client/datascience/jupyter/kernels/helpers.ts [hunk 3]
```diff
@@ -770,29 +769,42 @@ export async function sendTelemetryForPythonKernelExecutable(
...
         if (match) {
             return;
         }
 
+        // Rest of the code can all be async, no need to slow the calling code.
+
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/helpers.ts:778]: Comments, we don't need to be awaiting unnecessarily

File: src/client/datascience/jupyter/kernels/kernelCommandListener.ts [hunk 4]
```diff
@@ -180,12 +180,12 @@ export class KernelCommandListener implements IDataScienceCommandListener {
...
         try {
             await kernel.restart();
-            sendKernelTelemetryEvent(kernel.notebookDocument.uri, Telemetry.NotebookRestart, stopWatch.elapsedTime);
+            sendKernelTelemetryEvent(kernel.resourceUri, Telemetry.NotebookRestart, stopWatch.elapsedTime);
         } catch (exc) {
             // If we get a kernel promise failure, then restarting timed out. Just shutdown and restart the entire server.
             // Note, this code might not be necessary, as such an error is thrown only when interrupting a kernel times out.
             sendKernelTelemetryEvent(
-                kernel.notebookDocument.uri,
+                kernel.resourceUri,
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelCommandListener.ts:188]: Was passing the wrong Uri

File: src/client/datascience/jupyter/kernels/kernel.ts [hunk 5]
```diff
@@ -141,13 +138,13 @@ export class Kernel implements IKernel {
...
         );
-        trackKernelResourceInformation(notebookDocument.uri, {
+        trackKernelResourceInformation(resourceUri, {
             kernelConnection: kernelConnectionMetadata
         });
     }
     private perceivedJupyterStartupTelemetryCaptured?: boolean;
     public async executeCell(cell: NotebookCell): Promise<NotebookCellRunState> {
-        sendKernelTelemetryEvent(cell.notebook.uri, Telemetry.ExecuteCell);
+        sendKernelTelemetryEvent(this.resourceUri, Telemetry.ExecuteCell);
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:147]: With interacive window, we must use the `resourceUri`

File: src/client/datascience/jupyter/kernels/kernel.ts [hunk 6]
```diff
@@ -423,19 +444,23 @@ export class Kernel implements IKernel {
...
+            this._info.status !== 'ok'
+        ) {
             return;
         }
 
         const message = getSysInfoReasonHeader(reason, this.kernelConnectionMetadata);
-        const bannerMessage = (info.content as KernelMessage.IInfoReply)?.banner || '';
-        const sysInfoMessages = bannerMessage ? bannerMessage.split('\n') : [];
+        const sysInfoMessages = this._info.banner ? this._info.banner.split('\n') : [];
+        // TODO: This condition is wrong, it will always be true.
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:463]: Not entirely sure what this condition was
@joyceerhl ?

File: src/client/datascience/jupyter/kernels/kernel.ts [hunk 7]
```diff
@@ -248,13 +249,14 @@ export class Kernel implements IKernel {
                 const stopWatch = new StopWatch();
                 try {
                     try {
-                        const placeholderCell = await this.populateStartKernelInfoForInteractive(
+                        // No need to block kernel startup on UI updates.
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:252]: Perf improvement

File: src/client/datascience/jupyter/kernels/kernelCommandListener.ts [hunk 8]
```diff
@@ -194,8 +194,7 @@ export class KernelCommandListener implements IDataScienceCommandListener {
             if (exc instanceof JupyterKernelPromiseFailedError && kernel) {
                 // Old approach (INotebook is not exposed in IKernel, and INotebook will eventually go away).
                 const notebook = await this.notebookProvider.getOrCreateNotebook({
-                    resource: kernel.notebookDocument.uri,
-                    // TODO: This would be wrong for interactive window;
+                    resource: kernel.resourceUri,
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelCommandListener.ts:197]: Bug fix

File: src/client/datascience/jupyter/kernels/kernelDependencyService.ts [hunk 9]
```diff
@@ -127,9 +131,11 @@ export class KernelDependencyService implements IKernelDependencyService {
             ProductNames.get(Product.ipykernel)!
         );
         const ipykernelProductName = ProductNames.get(Product.ipykernel)!;
+        const resourceType = resource ? getResourceType(resource) : undefined;
         sendTelemetryEvent(Telemetry.PythonModuleInstal, undefined, {
             action: 'displayed',
-            moduleName: ipykernelProductName
+            moduleName: ipykernelProductName,
+            resourceType
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelDependencyService.ts:138]: For better telemetry

File: src/client/datascience/jupyter/kernels/kernelDependencyService.ts [hunk 10]
```diff
@@ -141,65 +147,82 @@ export class KernelDependencyService implements IKernelDependencyService {
...
-            sendTelemetryEvent(Telemetry.PythonModuleInstal, undefined, {
-                action: 'dismissed',
-                moduleName: ipykernelProductName
-            });
-            return KernelInterpreterDependencyResponse.cancel;
-        }
+        try {
+            if (!this.isCodeSpace) {
+                sendTelemetryEvent(Telemetry.PythonModuleInstal, undefined, {
+                    action: 'prompted',
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelDependencyService.ts:153]: For better telemetry

**Review by rchiodo — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**